File: src/main/java/MainWindows/MarvinPanel.java
package main.java.MainWindows;

import java.awt.Container;

/**
 * Main GUI class for Marvin integration with JBioFramework.
 * Instantiated as a singleton by /Main.JBioFrameworkMain/ as a JPanel object (extends JPanel).
 */

import java.awt.Dimension;
import java.util.function.Consumer;

import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.SwingUtilities;

import chemaxon.marvin.beans.MSketchPane;
import chemaxon.marvin.common.UserSettings;
import chemaxon.marvin.sketch.SketchParameterConstants;
import main.java.Utilities.JSUtil;

/**
 * JPanel object containing the main GUI for interacting with MarvinSketch in
 * Java or JavaScript.
 */
@SuppressWarnings("serial")
public class MarvinPanel extends JPanel {

	/**
	 * api id for allowed domains in JavaScript
	 * 
	 */
	private static String apiID = "2d2e4f98e04447babfbef7ae1e830756";

	private static int defaultWidth = 900;

	private static int defaultHeight = 500;

	/**
	 * the Marvin application, either Java or JavaScript
	 */
	private static MSketchPane marvinInstance;

	/**
	 * singleton instance of this class
	 */
	private static MarvinPanel instance;

	/**
	 * run after scripts are asynchronously added by jQuery
	 * 
	 */
	private static Runnable whenChemicalizeCreated;

	/**
	 * the name of the div we are using for Marvin; initially in document.body, but
	 * when the JFrame is added, this div is moved into it.
	 */
	private static String marvinDiv = "jbfmarvindiv";

	final static int STATE_UNITIALIZED = 0;
	final static int STATE_CHEMICALIZED = 1;
	final static int STATE_EDITORLOADED = 2;

	static int state = STATE_UNITIALIZED;

	private static JLabel labelWorking;

	private static Container marvinAppPanel;

	private static Object myclass = /**@j2sNative C$ || */null;

	private static void setupMarvinJS() {

		// JavaScript only

		// Add the div that Marvin JS will be placed into.

		// Add the Chemicalize scripts to the page,
		// running whenChemicalizeCreated only when ready.
		
		String serverScript = "https://marvinjs.chemicalize.com/v1/" + apiID + "/client-settings.js";
		String clientScript = "https://marvinjs.chemicalize.com/v1/client.js";
		String div = "<div style='display:none;width:" + defaultWidth + "px;height:" + defaultHeight + "px' id="
				+ marvinDiv + "></div>";
		System.out.println(serverScript + "\n" + clientScript + "\n" + div);

		@SuppressWarnings("unused")
		Runnable r = new Runnable() {

			@Override
			public void run() {
				state = STATE_CHEMICALIZED;
				if (whenChemicalizeCreated != null)
					whenChemicalizeCreated.run(); 
			}
			
		};

		/**
		 * @j2sNative
		 * 
		 * 			$("body").append(div);
		 */
		{
			// This block of code is never run, because it would only be in Java, but
			// this static code block is JavaScript-only. Just here as an illustration of
			// how the j2sNative annotation works.

		}

		// Using SwingUtilities.invokeLater is necessary to account for 
		// the fact that this static code is run before main() has had a chance to run.

		SwingUtilities.invokeLater(() -> {
			/**
			 * @j2sNative
			 * 
			 * 			$.getScript(serverScript, function() { $.getScript(clientScript, r.run$) });
			 */
		});

	}

	
	static {
		if (JSUtil.isJS) {
			setupMarvinJS();
		}

	}

	private MarvinPanel() {
		// private constructor only
	}

	/**
	 * Public access to a singleton instance.
	 * 
	 * @return the instance
	 */
	public static JPanel createMarvinPanel() {
		if (instance == null) {
			instance = new MarvinPanel();
			if (JSUtil.isJS)
				addPanel(null);
			else
				getMarvinApp(null);
		}
		return instance;
	}

	/**
	 * gets the Marvin application (asynchronously in JavaScript), running a
	 * Runnable when completed.
	 */
	public static void getMarvinApp(Runnable whenReady) {
		// BH 2021.04.26 allows lazy, asynchronous instantiation of Marvin for SwingJS
		if (marvinInstance != null) {
			if (whenReady != null)
				whenReady.run();
			return;
		}
		if (JSUtil.isJS) {
			createMarvinEditorJS(whenReady);
			return;
		}
		
		marvinInstance = new MSketchPane(createUserSettingsJava());
		addPanel(marvinInstance);

		if (whenReady != null)
			whenReady.run();
	}

	/**
	 * Create the JavaScript editor, addiing it to the frame when the
	 * ChemicalizeMarvinJs.createEditor promise is fulfilled.
	 * 
	 * This method only runs in JavaScript.
	 * 
	 * @param whenReady a Runnable to run when the editor is reports that it is
	 *                  ready
	 * 
	 * @author Bob Hanson hansonr@stolaf.edu
	 */
	@SuppressWarnings("unused") // some of these variables are JavaScript only
	private static void createMarvinEditorJS(Runnable whenReady) {
		
		transferMarvinDivIfYouCanJS();
		
		Consumer<Object> marvinReady = new Consumer<Object>() {

			@Override
			public void accept(Object marvinObject) {
				marvinInstance = (MSketchPane) marvinObject;
				if (labelWorking != null)
					labelWorking.setVisible(false);
				if (whenReady != null)
					whenReady.run();
			}

		};
		try {
			if (labelWorking != null)
				labelWorking.setText("loading Marvin JS...");
			/**
			 * @j2sNative
			 * 
			 * 			ChemicalizeMarvinJs.createEditor("#" +
			 *            C$.marvinDiv).then(function(m){marvinReady.accept$O(m)});
			 */
		} catch (Throwable e) {
			// Throwable includes all JavaScript Errors (as in Java),
			// which Exception does not.
			System.err.println(e);
		}
	}

	public void setVisible(boolean b) {
		super.setVisible(b);
		if (b) {
			getMarvinApp(null);
		}
	}
	
	private static void addPanel(Container marvin) {
		marvinAppPanel = (marvin == null ? new JPanel() : marvin);
		if (marvin == null)
			marvinAppPanel.add(labelWorking = new JLabel("working..."));
		marvinAppPanel.setPreferredSize(new Dimension(defaultWidth, defaultHeight));
		instance.add(marvinAppPanel);
		instance.validate();
		if (JSUtil.isJS) {
			transferMarvinDivIfYouCanJS();
		}
	}

	@SuppressWarnings("unused")
	private static void transferMarvinDivIfYouCanJS() {

		// Transfer the body-based marvinDiv into the frame
		// In SwingJS, the div associated with this panel is panel.ui.domNode 

		Container panel = marvinAppPanel;
		
		String mdiv = "#" + marvinDiv;
		/**
		 * @j2sNative 
		 * var d = $(panel.ui.domNode);
		 * var m = $(mdiv);
		 * d[0] && m.parent().is($("body")) && d.append(m.show());
		 */
		return;
	}

	/**
	 * The only method we are implementing here. Note that the JavaScript interface
	 * is different from the Java one.
	 * 
	 * @param mol
	 */
	public static void setMoleculeByName(String mol) {
		
		Object c = myclass ;
		getMarvinApp(() -> {
			/**
			 * @j2sNative 
			 * 
			 *   c.marvinInstance.importStructure("name", mol);
			 */
			{
				marvinInstance.setMol(mol);
			}
		});
	}

	/**
	 * Java only
	 * 
	 * Provide the back-end settings by using Marvin's classes.. construct and
	 * return Marvin's /UserSettings/ object which is used in the construction of
	 * 'pane' above.
	 */
	private static UserSettings createUserSettingsJava() {
		UserSettings settings = new UserSettings(MarvinPanel.class.getResourceAsStream("marvin.properties"));
		settings.setProperty(SketchParameterConstants.MENU_CUSTOMIZATION_FILE,
				System.getProperty("user.dir") + "/src/customization.xml");
		settings.setProperty(SketchParameterConstants.SHORTCUTS, System.getProperty("user.dir") + "/src/shortcuts.xml");
		settings.setProperty(SketchParameterConstants.TOOLBAR_TEMPLATES + "20", ":specials:specialTemplates.mrv");
		return settings;
	}

	/**
	 * Create a JFrame that just contains the Marvin app for testing.
	 * 
	 * Note that this will not actually work as an app in JavaScript.
	 * 
	 * @param args
	 */
	public static void main(String[] args) {
		whenChemicalizeCreated = new Runnable() {
			@Override
			public void run() {
				whenChemicalizeCreated = null;
				JFrame frame = new JFrame();
				frame.add(createMarvinPanel());
				frame.pack();					
				frame.setVisible(true);
				getMarvinApp(() -> {
					setMoleculeByName("Glycyl-Alanine");
				}); 
			}

		};
		if (!JSUtil.isJS || state == STATE_CHEMICALIZED)
			whenChemicalizeCreated.run();
	}
}

File: site/swingjs/j2s/main/java/MainWindows/MarvinPanel.js
(function(){var P$=Clazz.newPackage("main.java.MainWindows"),I$=[[0,'main.java.Utilities.JSUtil','main.java.MainWindows.MarvinPanel','javax.swing.SwingUtilities','chemaxon.marvin.beans.MSketchPane','javax.swing.JPanel','javax.swing.JLabel','java.awt.Dimension','chemaxon.marvin.common.UserSettings','javax.swing.JFrame']],I$0=I$[0],$I$=function(i,n,m){return m?$I$(i)[n].apply(null,m):((i=(I$[i]||(I$[i]=Clazz.load(I$0[i])))),!n&&i.$load$&&Clazz.load(i,2),i)};
/*c*/var C$=Clazz.newClass(P$, "MarvinPanel", null, 'javax.swing.JPanel');

C$.$clinit$=2;

Clazz.newMeth(C$, '$init$', function () {
},1);

C$.$fields$=[[]
,['I',['defaultWidth','defaultHeight','state'],'S',['apiID','marvinDiv'],'O',['marvinInstance','chemaxon.marvin.beans.MSketchPane','instance','main.java.MainWindows.MarvinPanel','whenChemicalizeCreated','Runnable','labelWorking','javax.swing.JLabel','marvinAppPanel','java.awt.Container','myclass','java.lang.Object']]]

Clazz.newMeth(C$, 'setupMarvinJS$',  function () {
var serverScript="https://marvinjs.chemicalize.com/v1/" + C$.apiID + "/client-settings.js" ;
var clientScript="https://marvinjs.chemicalize.com/v1/client.js";
var div="<div style='display:none;width:" + C$.defaultWidth + "px;height:" + C$.defaultHeight + "px' id=" + C$.marvinDiv + "></div>" ;
System.out.println$S(serverScript + "\n" + clientScript + "\n" + div );
var r=((P$.MarvinPanel$1||
(function(){/*a*/var C$=Clazz.newClass(P$, "MarvinPanel$1", function(){Clazz.newInstance(this, arguments[0],1,C$);}, null, 'Runnable', 1);

C$.$clinit$=2;

Clazz.newMeth(C$, '$init$', function () {
},1);

Clazz.newMeth(C$, 'run$',  function () {
$I$(2).state=1;
if ($I$(2).whenChemicalizeCreated != null ) $I$(2).whenChemicalizeCreated.run$();
});
})()
), Clazz.new_(P$.MarvinPanel$1.$init$,[this, null]));
{
$("body").append(div);
}
$I$(3,"invokeLater$Runnable",[(P$.MarvinPanel$lambda1$||(P$.MarvinPanel$lambda1$=(((P$.MarvinPanel$lambda1||
(function(){/*m*/var C$=Clazz.newClass(P$, "MarvinPanel$lambda1", function(){Clazz.newInstance(this, arguments[0],1,C$);}, null, 'Runnable', 1);

C$.$clinit$=2;

Clazz.newMeth(C$, '$init$', function () {
},1);
/*lambda_E*/
Clazz.newMeth(C$, 'run$',  function () {

$.getScript(serverScript, function() { $.getScript(clientScript, r.run$) });
});
})()
), Clazz.new_(P$.MarvinPanel$lambda1.$init$,[this, null])))))]);
}, 1);

Clazz.newMeth(C$, 'c$',  function () {
Clazz.super_(C$, this);
}, 1);

Clazz.newMeth(C$, 'createMarvinPanel$',  function () {
if (C$.instance == null ) {
C$.instance=Clazz.new_(C$);
if ($I$(1).isJS) C$.addPanel$java_awt_Container(null);
 else C$.getMarvinApp$Runnable(null);
}return C$.instance;
}, 1);

Clazz.newMeth(C$, 'getMarvinApp$Runnable',  function (whenReady) {
if (C$.marvinInstance != null ) {
if (whenReady != null ) whenReady.run$();
return;
}if ($I$(1).isJS) {
C$.createMarvinEditorJS$Runnable(whenReady);
return;
}C$.marvinInstance=Clazz.new_([C$.createUserSettingsJava$()],$I$(4,1).c$$chemaxon_marvin_common_UserSettings);
C$.addPanel$java_awt_Container(C$.marvinInstance);
if (whenReady != null ) whenReady.run$();
}, 1);

Clazz.newMeth(C$, 'createMarvinEditorJS$Runnable',  function (whenReady) {
C$.transferMarvinDivIfYouCanJS$();
var marvinReady=((P$.MarvinPanel$2||
(function(){/*a*/var C$=Clazz.newClass(P$, "MarvinPanel$2", function(){Clazz.newInstance(this, arguments[0],1,C$);}, null, 'java.util.function.Consumer', 1);

C$.$clinit$=2;

Clazz.newMeth(C$, '$init$', function () {
},1);

Clazz.newMeth(C$, 'accept$O',  function (marvinObject) {
$I$(2).marvinInstance=marvinObject;
if ($I$(2).labelWorking != null ) $I$(2).labelWorking.setVisible$Z(false);
if (this.$finals$.whenReady != null ) this.$finals$.whenReady.run$();
});
})()
), Clazz.new_(P$.MarvinPanel$2.$init$,[this, {whenReady:whenReady}]));
try {
if (C$.labelWorking != null ) C$.labelWorking.setText$S("loading Marvin JS...");

ChemicalizeMarvinJs.createEditor("#" + C$.marvinDiv).then(function(m){marvinReady.accept$O(m)});
} catch (e) {
System.err.println$O(e);
}
}, 1);

Clazz.newMeth(C$, 'setVisible$Z',  function (b) {
C$.superclazz.prototype.setVisible$Z.apply(this, [b]);
if (b) {
C$.getMarvinApp$Runnable(null);
}});

Clazz.newMeth(C$, 'addPanel$java_awt_Container',  function (marvin) {
C$.marvinAppPanel=(marvin == null  ? Clazz.new_($I$(5,1)) : marvin);
if (marvin == null ) C$.marvinAppPanel.add$java_awt_Component(C$.labelWorking=Clazz.new_($I$(6,1).c$$S,["working..."]));
C$.marvinAppPanel.setPreferredSize$java_awt_Dimension(Clazz.new_($I$(7,1).c$$I$I,[C$.defaultWidth, C$.defaultHeight]));
C$.instance.add$java_awt_Component(C$.marvinAppPanel);
C$.instance.validate$();
if ($I$(1).isJS) {
C$.transferMarvinDivIfYouCanJS$();
}}, 1);

Clazz.newMeth(C$, 'transferMarvinDivIfYouCanJS$',  function () {
var panel=C$.marvinAppPanel;
var mdiv="#" + C$.marvinDiv;

var d = $(panel.ui.domNode);
var m = $(mdiv);
d[0] && m.parent().is($("body")) && d.append(m.show());
return;
}, 1);

Clazz.newMeth(C$, 'setMoleculeByName$S',  function (mol) {
var c=C$.myclass;
C$.getMarvinApp$Runnable((P$.MarvinPanel$lambda2$||(P$.MarvinPanel$lambda2$=(((P$.MarvinPanel$lambda2||
(function(){/*m*/var C$=Clazz.newClass(P$, "MarvinPanel$lambda2", function(){Clazz.newInstance(this, arguments[0],1,C$);}, null, 'Runnable', 1);

C$.$clinit$=2;

Clazz.newMeth(C$, '$init$', function () {
},1);
/*lambda_E*/
Clazz.newMeth(C$, 'run$',  function () {
{
c.marvinInstance.importStructure("name", mol);
}
});
})()
), Clazz.new_(P$.MarvinPanel$lambda2.$init$,[this, null]))))));
}, 1);

Clazz.newMeth(C$, 'createUserSettingsJava$',  function () {
var settings=Clazz.new_([Clazz.getClass(C$).getResourceAsStream$S("marvin.properties")],$I$(8,1).c$$java_io_InputStream);
settings.setProperty$S$S("menuconfig", System.getProperty$S("user.dir") + "/src/customization.xml");
settings.setProperty$S$S("shortcuts", System.getProperty$S("user.dir") + "/src/shortcuts.xml");
settings.setProperty$S$S("ttmpls20", ":specials:specialTemplates.mrv");
return settings;
}, 1);

Clazz.newMeth(C$, 'main$SA',  function (args) {
C$.whenChemicalizeCreated=((P$.MarvinPanel$3||
(function(){/*a*/var C$=Clazz.newClass(P$, "MarvinPanel$3", function(){Clazz.newInstance(this, arguments[0],1,C$);}, null, 'Runnable', 1);

C$.$clinit$=2;

Clazz.newMeth(C$, '$init$', function () {
},1);

Clazz.newMeth(C$, 'run$',  function () {
$I$(2).whenChemicalizeCreated=null;
var frame=Clazz.new_($I$(9,1));
frame.add$java_awt_Component($I$(2).createMarvinPanel$());
frame.pack$();
frame.setVisible$Z(true);
$I$(2,"getMarvinApp$Runnable",[(P$.MarvinPanel$3$lambda3$||(P$.MarvinPanel$3$lambda3$=(((P$.MarvinPanel$3$lambda3||
(function(){/*m*/var C$=Clazz.newClass(P$, "MarvinPanel$3$lambda3", function(){Clazz.newInstance(this, arguments[0],1,C$);}, null, 'Runnable', 1);

C$.$clinit$=2;

Clazz.newMeth(C$, '$init$', function () {
},1);
/*lambda_E*/
Clazz.newMeth(C$, 'run$',  function () {
$I$(2).setMoleculeByName$S("Glycyl-Alanine");
});
})()
), Clazz.new_(P$.MarvinPanel$3$lambda3.$init$,[this, null])))))]);
});
})()
), Clazz.new_(P$.MarvinPanel$3.$init$,[this, null]));
if (!$I$(1).isJS || C$.state == 1 ) C$.whenChemicalizeCreated.run$();
}, 1);

C$.$static$=function(){C$.$static$=0;
C$.apiID="2d2e4f98e04447babfbef7ae1e830756";
C$.defaultWidth=900;
C$.defaultHeight=500;
C$.marvinDiv="jbfmarvindiv";
C$.state=0;
C$.myclass=C$ ||null;
{
if ($I$(1).isJS) {
C$.setupMarvinJS$();
}};
};
})();
;Clazz.setTVer('3.3.1-v4');//Created 2022-04-21 16:53:54 Java2ScriptVisitor version 3.3.1-v4 net.sf.j2s.core.jar version 3.3.1-v4
